src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -640,127 +465,54 @@ private void VerifyCanCreateOperationResponseMessage(bool synchronousCall)
-
-            this.SetState(BatchWriterState.OperationStreamRequested);
-        }
-
-        /// <summary>
-        /// Remember a non-null Content-ID header for change set request operations.
-        /// If a non-null content ID header is specified for a change set request operation, record it in the URL resolver.
-        /// </summary>
-        /// <param name="contentId">The Content-ID header value read from the message.</param>
-        /// <remarks>
-        /// Note that the content ID of this operation will only
-        /// become visible once this operation has been written
-        /// and OperationCompleted has been called on the URL resolver.
-        /// </remarks>
-        private void RememberContentIdHeader(string contentId)
-        {
-            // The Content-ID header is only supported in request messages and inside of changesets.
-            Debug.Assert(this.currentOperationRequestMessage != null, "this.currentOperationRequestMessage != null");
-            Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
-
-            // Set the current content ID. If no Content-ID header is found in the message,
-            // the 'contentId' argument will be null and this will reset the current operation content ID field.
-            this.currentOperationContentId = contentId;
-
-            // Check for duplicate content IDs; we have to do this here instead of in the cache itself
-            // since the content ID of the last operation never gets added to the cache but we still
-            // want to fail on the duplicate.
-            if (contentId != null && this.urlResolver.ContainsContentId(contentId))
-            {
-                throw new ODataException(Strings.ODataBatchWriter_DuplicateContentIDsNotAllowed(contentId));
-            }
-        }
+        protected abstract void DisposeBatchWriterAndSetContentStreamRequestedState();
 
         /// <summary>
         /// Verifies that the writer is in correct state for the Flush operation.
         /// </summary>
         /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCanFlush(bool synchronousCall)
-        {
-            this.rawOutputContext.VerifyNotDisposed();
-            this.VerifyCallAllowed(synchronousCall);
-            if (this.state == BatchWriterState.OperationStreamRequested)
-            {
-                this.ThrowODataException(Strings.ODataBatchWriter_FlushOrFlushAsyncCalledInStreamRequestedState);
-            }
-        }
+        protected abstract void VerifyCanFlush(bool synchronousCall);
 
         /// <summary>
         /// Verifies that a call is allowed to the writer.
         /// </summary>
         /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCallAllowed(bool synchronousCall)
+        internal void VerifyCallAllowed(bool synchronousCall)

```

Comment:
>internal [](start = 8, length = 8)

Should be able to make this private once you restore VerifyCanFlush to a private method in this class.